src/nvim/channel.h

```diff
@@ -0,0 +1,138 @@
+#ifndef NVIM_CHANNEL_H
+#define NVIM_CHANNEL_H
+
+#include "nvim/main.h"
+#include "nvim/event/socket.h"
+#include "nvim/event/process.h"
+#include "nvim/os/pty_process.h"
+#include "nvim/event/libuv_process.h"
+#include "nvim/eval/typval.h"
+#include "nvim/msgpack_rpc/channel_defs.h"
+
+#define CHANNEL_BUFFER_SIZE 0xffff
+
+typedef enum {
+  kChannelTypeProc,
+  kChannelTypeSocket,
+  kChannelTypeStdio,
+  kChannelTypeStderr,
+  kChannelTypeInternal
+} ChannelType;
+
+typedef enum {
+  kChannelPartStdin,
+  kChannelPartStdout,
+  kChannelPartStderr,
+  kChannelPartRpc,
+  kChannelPartAll
+} ChannelPart;
+
+
+typedef struct {
+  Stream in;
+  Stream out;
+} StdioPair;
+
+typedef struct {
+  bool closed;
+} StderrState;
+
+typedef struct {
+  Callback cb;
+  garray_T buffer;
+  bool buffered;
+} CallbackReader;
+
+#define CALLBACK_READER_INIT ((CallbackReader){ .cb = CALLBACK_NONE,  \
+                                                .buffer = {0},        \
+                                                .buffered = false })
+static inline bool callback_reader_set(CallbackReader reader)
+{
+  return reader.cb.type != kCallbackNone;
+}
+
+
+struct Channel {
+  uint64_t id;
+  size_t refcount;
+  ChannelType type;
+  union {
+    Process proc; // FIXME: very UB?
+    LibuvProcess uv;
+    PtyProcess pty;
+    Stream socket;
+    StdioPair stdio;
+    StderrState stderr;
+  } stream;
+
+  //ProtocolType protocol;
+  bool is_rpc;
+  //union {
+    Terminal* term;
+    RpcState rpc;
+  //};
+
+  CallbackReader on_stdout;
+  CallbackReader on_stderr;
+  Callback on_exit;
+  varnumber_T *status_ptr; // TODO: refactor?
+  MultiQueue *events;
+};
+
+EXTERN PMap(uint64_t) *channels;
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "channel.h.generated.h"
+#endif
+
+static inline Channel *find_channel(uint64_t id)
+{
+  Channel *data = pmap_get(uint64_t)(channels, id);
+  if (!data) {
```

Comment:
The whole function body could be replaced with `return pmap_get(uint64_t)(channels, id);`